### PR TITLE
[RSDK-7813] add FakeCamera

### PR DESCRIPTION
**Pull request by mattjperez**
These changes let you register a `FakeCamera` on app.viam and be able to connect to it via SDK.

still TODO:
- [x] Add a static image to `FakeCamera` for verifying `GetImage`
- [x] Confirm image grabbed with GetImage on SDK over HTTP2
  - confirmed on go and python

Ended up going a little further than originally planned for this PR. It enables (for native-only):
- GetImage
- ~RenderFrame (mimics GetImage, displays on app.viam)~ deprecated except for RC page. will maybe revisit in later ticket/PR
- DoCommand

### CODE REVIEW

File: micro-rdk-server/Cargo.toml
Comment (mattjperez) [inline micro-rdk-server/Cargo.toml:20]: added for development convenience, remove before PR is merged

File: micro-rdk/src/common/robot.rs
Comment (mattjperez) [inline micro-rdk/src/common/robot.rs:191]: Neither the `RUST_LOG=debug` errors nor app.viam logs were helpful when troubleshooting init failure for camera. Not directly related to the ticket but may be useful to add either here or separate PR

File: Makefile
Comment (mattjperez) [inline Makefile:40]: also for development convenience. used to bypass provisioning and use the local `viam.json`. 
will remove before merging

File: micro-rdk/src/common/camera.rs
Comment (npmenard) [inline micro-rdk/src/common/camera.rs:40]: should not be `unimplemented!()` either without a body or returning an error

File: micro-rdk/src/common/camera.rs
Comment (npmenard) [inline micro-rdk/src/common/camera.rs:69]: should be removed once done with dev

File: micro-rdk/src/common/grpc.rs
Comment (npmenard) [inline micro-rdk/src/common/grpc.rs:1255]: if theses functions (including point cloud, render frame etc.. are going to remain unimplemented we should remove the code because by default we return unimplemented on all unknown RPCs

File: micro-rdk/src/common/grpc.rs
Comment (npmenard) [inline micro-rdk/src/common/grpc.rs:221]: rename frame->image

File: micro-rdk/src/common/camera.rs
Comment (npmenard) [inline micro-rdk/src/common/camera.rs:38]: rename frame->image

File: micro-rdk/src/common/robot.rs
Comment (npmenard) [reply micro-rdk/src/common/robot.rs:191]: format the error reporting better in this case so there is a bit more context

**Review by npmenard — CHANGES_REQUESTED** [review]
done with first pass

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [reply micro-rdk/src/common/grpc.rs:1255]: So remove unimplemented rpcs from `grpc.rs` but leave the trait methods in `camera.rs` (returning an error/leaving them blank)?

File: micro-rdk/src/common/webrtc/grpc.rs
Comment (mattjperez) [inline micro-rdk/src/common/webrtc/grpc.rs:96]: @npmenard I used the Python SDK with `disable_webrtc=True` but the native server still went down this path with webrtc. I need to try again with the Go SDK, but this change did allow me to get everything working.

Not sure what this value should be if it's the limiting factor. We could feature-gate whatever the value is as well.

File: micro-rdk/src/lib.rs
Comment (mattjperez) [inline micro-rdk/src/lib.rs:29]: needed for `google.api.HttpBody`, the response type for `RenderFrame`

**Comment by mattjperez** [comment]
Ended up going a little further than originally planned for this PR. It enables:
- GetImage
- RenderFrame
- DoCommand

These should be everything I need for e2e testing for the rest of the epic.

File: micro-rdk/src/lib.rs
Comment (mattjperez) [inline micro-rdk/src/lib.rs:32]: for `RenderFrame`'s `HttpBody` response

File: micro-rdk/src/common/camera.rs
Comment (mattjperez) [inline micro-rdk/src/common/camera.rs:88]: Will change these to return errors

File: micro-rdk/src/common/camera.rs
Comment (npmenard) [inline micro-rdk/src/common/camera.rs:105]: remove the unwrap and return an error

File: micro-rdk/src/common/grpc.rs
Comment (npmenard) [inline micro-rdk/src/common/grpc.rs:1259]: remove comments

File: micro-rdk/src/common/camera.rs
Comment (npmenard) [inline micro-rdk/src/common/camera.rs:101]: do we make any use of HTTP extensions?

**Review by npmenard — CHANGES_REQUESTED** [review]
Done with Pass! Nice work

File: micro-rdk/src/common/camera.rs
Comment (mattjperez) [reply micro-rdk/src/common/camera.rs:101]: not in [rdk camera server](https://github.com/viamrobotics/rdk/blob/04037e3ddc9abd2740b1e3d5f3e8152e04e3b121/components/camera/server.go#L194) at least

File: micro-rdk/src/common/camera.rs
Comment (acmorrow) [inline micro-rdk/src/common/camera.rs:19]: Any reason not to just put this in the same directory as `camera.rs`, or under an `assets` directory there rather than at the top?

And `symbol` is a pretty generic name. Maybe `fake_camera_image`?.

Comment (mattjperez) [reply micro-rdk/src/common/camera.rs:19]: I was avoiding putting artifacts directly in the micro-rdk core repo.
fine with putting it here though. 

I'm tempted to move the FakeCamera and the image into the native directory and leave just the trait definition here. 

As for the name, I had several candidate files when testing, (symbol, checkers, etc). will rename it

Comment (acmorrow) [reply micro-rdk/src/common/camera.rs:19]: Yes, it should live next to whatever ends up using it, wherever that lives.

File: micro-rdk/src/lib.rs
Comment (mattjperez) [reply micro-rdk/src/lib.rs]: render frame no longer being implemented this PR

**Review by npmenard — APPROVED** [review]
LGTM pending tests over HTTP2

File: micro-rdk/src/common/registry.rs
Comment (gvaradarajan) [inline micro-rdk/src/common/registry.rs:168]: `fake_camera` only exists if the "native" feature is activated, but here we attempt to register models even if the native feature is not active. Your PR seems to be passing CI checks despite this but I'm not sure how?

EDIT: it's because we don't pass the camera feature in any of our CI checks (which I'm not necessarily advocating for doing). I've confirmed that the following fails
```
cargo +esp clippy -p micro-rdk  --features esp32,provisioning,camera  --target=xtensa-esp32-espidf -Zbuild-std=std,panic_abort -- -Dwarnings
```
The follow-up question here is: should we seek a more graceful compilation failure via [compile_error](https://doc.rust-lang.org/std/macro.compile_error.html)?

**Review by gvaradarajan — CHANGES_REQUESTED** [review]
LGTM, but I would like the one thing I noticed addressed somehow (or a decision made on it in consultation with @npmenard and @acmorrow )

File: micro-rdk/src/common/registry.rs
Comment (mattjperez) [reply micro-rdk/src/common/registry.rs:168]: For the `Esp32Camera`, this ticket only covers native. In the [epic](https://viam.atlassian.net/browse/RSDK-7813?atlOrigin=eyJpIjoiYzFmMWNkNmZjNzkxNDVmMjgxNzY0MDNiNTllZWMyMDIiLCJwIjoiaiJ9), esp32 implementations and testings come later.

Comment (mattjperez) [reply micro-rdk/src/common/registry.rs:168]: I can do an immediate PR after so it passes for esp32,camera. Or if you don't mind waiting a week and I fix it when I get to the esp32 stuff

Comment (gvaradarajan) [reply micro-rdk/src/common/registry.rs:168]: Could you put the native flag above this line though? Otherwise even when you fix `Esp32Camera` it will still fail

**Review by gvaradarajan — APPROVED** [review]
LGTM

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [inline micro-rdk/src/common/grpc.rs:38]: grpc spec has an incoming message limit of 4MB. the current `fake_image.jpeg` is about 10KB, the rdk's `FakeCamera` saves a jpeg of ~20KB.

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [reply micro-rdk/src/common/grpc.rs:1236]: Should I change the line in [`encode_message`](https://github.com/viamrobotics/micro-rdk/blob/26dc4a70874461f5679d76356afd2b186668aaf2/micro-rdk/src/common/grpc.rs#L1279) as well for consistency?

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [reply micro-rdk/src/common/grpc.rs:1234]: for example

```rs
    let mut first = BytesMut::with_capacity(9000);
    println!("first capacity: {}", first.capacity());

    let second = first.split_off(100);
    println!("first capacity: {}", first.capacity());
    println!("second capacity: {}", second.capacity());
```

output:
```sh
first capacity: 9000
first capacity: 100
second capacity: 8900
```

Comment (npmenard) [reply micro-rdk/src/common/grpc.rs:1234]: yes seems logical if you split the buffer with split_off. left would have [0,100) and right [100,9000) so capacity returned is correct

File: micro-rdk/src/common/grpc.rs
Comment (npmenard) [reply micro-rdk/src/common/grpc.rs:1236]: In encode_message we check if data will fit before we do the actual encoding. (hence the call tne encoded_len)

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [reply micro-rdk/src/common/grpc.rs:1234]: since I'm removing the following check this comment will go as well.

**Pull request merged**
